### DISCUSSION TITLE
How do you use Znai?

### DISCUSSION THREAD

**Discussion by MykolaGolubyev** [testingisdocumenting/znai#824] [Show and tell]
Hello,
I am very curios to know how do you use Znai.
Your use cases will greatly help to prioritize existing and every growing backlog!

**Comment by adzubla** [comment]
I'm using znai to create the documentation of an internal project, creating the static files with `mvn znai:build`

The generated files are uploaded to AWS S3 and they are exposed by AWS CloudFront.

As this content is not public, I added authentication using this:
https://aws.amazon.com/pt/blogs/networking-and-content-delivery/authorizationedge-using-cookies-protect-your-amazon-cloudfront-content-from-being-downloaded-by-unauthenticated-users/

Previously, we had a lot of MarkDown documents in Stoplight, but it was fairly easy to migrate them to znai.

Congrats for the nice work!

**Reply by MykolaGolubyev** [reply]
Thank you for sharing!
Once your documentation matures, I would be curios to see what plugins do you use. 
Znai generates usage-stats json file. If you are comfortable, I would like to take a look at it. Or you can mention top plugins you find useful.

**Reply by MykolaGolubyev** [reply]
Just in case, if you are documenting REST API, take a look at https://testingisdocumenting.org/znai/synergy-with-testing/REST-API if you haven't yet

**Reply by MykolaGolubyev** [reply]
@adzubla wanted to check if you need any help

**Comment by adzubla** [comment]
Hi,

thats the `plugin-stats.json` generated by znai:build

[plugin-stats.txt](https://github.com/testingisdocumenting/znai/files/9162218/plugin-stats.txt)

**Reply by MykolaGolubyev** [reply]
Thank you. Seems like for now you are only using images. Let me know if you need help with any other plugins in znai

**Comment by Haxiz** [comment]
We are using Znai for documenting our Discord Bot [Mantaro](https://github.com/Mantaro/MantaroBot).

Before Znai we used Github's wiki option for out documentation, however that is a bit limited compared to Znai.
The switch over was brought by a big update for the bot coming soon, and the Github wiki being outdated.

Our wiki looks to document as much as we can about the bot so that anyone would be able to find whatever information they need, and be able to understand it without being too technical (on some things its unavoidable).
This goes from documenting our all our commands, badges and items. Writing tutorials and examples for people to easily understand some of the more complex commands, as well as our currency system, and so on. In general, being able to provide as much helpful information we can to our users.

The current Github wiki can be found [here](https://github.com/Mantaro/MantaroBot/wiki), thought it might be gone in a few days at the time of writing this.
As well, our wiki with Znai can be found [here](https://www.mantaro.site/mantaro-wiki/). Once again, at the time of writing this it is still a work in progress. Some pages are still just copy paste from our old Github wiki, some are done and just haven't been deployed yet. (Hopefully i'll have it done before september).

**Reply by MykolaGolubyev** [reply]
Thank you for the very detailed answer! Let me know if you want me to review docs at some point and maybe suggest znai plugins to streamline the result

**Reply by MykolaGolubyev** [reply]
Just looked at your table with icons here https://www.mantaro.site/mantaro-wiki/commands/permissions#intro
Nice! Glad you found it. Funny enough table icons support was added for our internal wiki migration :)

**Reply by Haxiz** [reply]
Yes! I really liked the way they look and i'll probably use it in other places. The thing we're looking into right now is moving stuff over to templates. I found them way too late and wrote way too much by hand haha so now I want to move it over.
As well, I was looking for alternatives for tables. Things like https://www.mantaro.site/mantaro-wiki/currency/badges use a lot of them, and looking into the mobile view i noticed that tables tend to be scrollable horizontally. So doing a lot of info probably wouldnt be great on a table.
I was thinking maybe we could use columns to show image on one side, and text on the other. If you have any suggestions for that I would be glad to hear them!

**Reply by MykolaGolubyev** [reply]
Good point on mobile view and scrolls. Columns in mobile view with image may work. However it may also become stacked instead, depending on width. 

One way is to use templates to define a single entry and then make use of columns inside. Or even template loop and json data (I may have deleted that doc page inside templates). 
another is to use api parameters: parameter name would be badge name, description, etc

I believe I have a card component in a backlog but never had a good use case. Maybe a card with an image in the corner and free form text may work.

**Reply by Haxiz** [reply]
Interesting options! Yeah cards would be a godsend if they were an option, but i can see why they dont really have a lot of use cases. I can already see myself trying to merge current options with cards haha having a column inside a column to effectively have 3 columns, then placing 3 cards on it is what comes to my mind haha
Trying to work with what we have right now i think maybe working with columns for now. I feel like we spammed api params so much on other pages already haha so a change in look would be good. Id just have to find a good balance with the image size and so on so its not too stuffy to look at, and have it still work on both mobile and pc.

**Reply by MykolaGolubyev** [reply]
Fair point on api params. 
any chance you can sketch the look and feel you have in mind?
I am up to adding new reusable things!

**Reply by Haxiz** [reply]
Well, im not that much of a sketcher haha specially when not on my pc
Buut heres a small very rough drawing in paint of what my mind came up with

and when i say rough i mean **very rough**

![image](https://user-images.githubusercontent.com/31944205/187323569-0492475e-23d2-45ae-aaaa-ddc4cfe507fa.png)

**Reply by MykolaGolubyev** [reply]
This is pretty good. If you were to pick one as a starter, which one would it be?

**Reply by Haxiz** [reply]
probably a mix between the first and two, or the first one with the title option, the horizontal one sounds like a cool idea but i probably see myself using the vertical one more

**Reply by MykolaGolubyev** [reply]
<img width="477" alt="Screen Shot 2022-09-10 at 9 35 12 PM" src="https://user-images.githubusercontent.com/11657860/189507893-c73efe99-8c59-4b56-b475-8dca30cb5ec9.png">

**Reply by Haxiz** [reply]
It looks great!
I love that you knew I would use api params on it haha

**Reply by MykolaGolubyev** [reply]
Thanks!
Do you plan to use small or large images? Right now small image will be centered and if it has background it may look weird. Large image will be shrieked down to fit. 
I was thinking to add an option to set an image background color for small ones. Or an option to stretch (don’t like it really).

**Reply by Haxiz** [reply]
Mostly small images i think, something like our [item page](https://www.mantaro.site/mantaro-wiki/currency/items) will probably use it a lot due to the sheer amount of images it should have.

An item example would be:
![image](https://user-images.githubusercontent.com/31944205/189534803-ae4b2405-1ad7-41f9-b941-61bbe16ad3eb.png)

We have some images as pngs, and others as svgs, I wonder if that would affect it somehow? I can always get them all on the same format (which i probably will anyway, but good to know anyway)

> Right now small image will be centered and if it has background it may look weird.
Or an option to stretch (don’t like it really).

Maybe instead of trying to fit the image to the card, fit the card to the image size? Though I'm not sure how well that could work with the text. Maybe an option for the card to have a set width?

**Reply by MykolaGolubyev** [reply]
Thanks for the info. Card size based on image may not work well. Right now I set the max size of the card to be half of the full width. In three columns mode it will be shrank accordingly. 

Prefer not to let users set the width as it may not play well with different views. 

Will have to check if svg works. 
let me grab one of your images and play around. 

Also we can move this discussion away now as a ticket if you don’t mind creating one.

**Comment by jamesjreynolds** [comment]
We are using Znai to produce client facing documentation of our APIs and SDKs.  Using Github actions, it's a simple matter to build and deploy to s3 and distribute over a CDN.

Thanks for a great product!
